app/managers/network_performance_manager.ts

```diff
@@ -0,0 +1,325 @@
+     */
+    public startRequestTracking = (serverUrl: string, url: string): string => {
+        const requestId = `${Date.now()}-${Math.random().toString(36).slice(2, 11)}`;
+        if (!this.activeRequests[serverUrl]) {
+            this.activeRequests[serverUrl] = {};
+        }
+
+        const request: ActiveRequest = {
+            id: requestId,
+            url,
+            startTime: Date.now(),
+        };
+
+        request.checkTimer = setTimeout(() => {
+            this.checkRequestLatency(serverUrl, requestId);
+        }, SLOW_REQUEST_THRESHOLD);
+
+        this.activeRequests[serverUrl][requestId] = request;
+        return requestId;
+    };
+
+    /**
+     * Completes request tracking and adds metrics for performance reporting.
+     * Should be called when the request finishes with the ID from startRequestTracking.
+     */
+    public completeRequestTracking = (serverUrl: string, requestId: string, metrics: ClientResponseMetrics | undefined) => {
+        const activeRequest = this.activeRequests[serverUrl]?.[requestId];
+        const wasEarlyDetected = !activeRequest; // If not found, it was already early detected and removed
+
+        this.clearActiveRequest(serverUrl, requestId);
+
+        // Only record the outcome if it wasn't already recorded by early detection
+        if (!wasEarlyDetected && metrics?.latency) {
+            this.recordRequestOutcome(serverUrl, {
+                timestamp: Date.now(),
+                isSlow: metrics.latency >= SLOW_REQUEST_THRESHOLD,
+                wasEarlyDetection: false,
+            });
+        }
+    };
+
+    /**
+     * Cancels request tracking when a request fails.
+     * Should be called when the request fails with the ID from startRequestTracking.
+     */
+    public cancelRequestTracking = (serverUrl: string, requestId: string) => {
+        this.clearActiveRequest(serverUrl, requestId);
+    };
+
+    /**
+     * Returns an observable that emits network performance state changes.
+     * Emits 'normal' or 'slow' based on current performance metrics.
+     */
+    public observePerformanceState = (serverUrl: string) => {
+        return this.getPerformanceSubject(serverUrl).asObservable().pipe(
+            distinctUntilChanged(),
+        );
+    };
+
+    /**
+     * Gets the current network performance state for a server.
+     */
+    public getCurrentPerformanceState = (serverUrl: string): NetworkPerformanceState => {
+        return this.getPerformanceSubject(serverUrl).getValue();
+    };
+
+    /**
+     * Gets the current request outcome statistics for a server.
+     */
+    public getRequestOutcomeStats = (serverUrl: string) => {
+        const outcomes = this.requestOutcomes[serverUrl] || [];
+        if (!outcomes.length) {
+            return {
+                totalRequests: 0,
+                slowRequests: 0,
+                slowPercentage: 0,
+                earlyDetectionCount: 0,
+            };
+        }
+
```

Comment:
I'm wondering if it should also have a filter of queries being too old. A request that failed 15 min ago (not sure where the threshold should be) shouldn't affect the current state of the connection.